server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -1695,8 +1697,25 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
     /**
      * Returns a new {@link QueryRewriteContext} with the given {@code now} provider
      */
-    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis) {
-        return new QueryRewriteContext(parserConfig, client, nowInMillis);
+    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis, IndicesRequest indicesRequest) {
+        // Skip unavailable indices
+        // TODO: Support metadata lookups for indices on remote clusters
+        IndicesOptions.Builder indicesOptionsBuilder = IndicesOptions.builder(indicesRequest.indicesOptions())
+            .concreteTargetOptions(new IndicesOptions.ConcreteTargetOptions(true));
+        Index[] indices = indexNameExpressionResolver.concreteIndices(
+            clusterService.state(),
+            indicesOptionsBuilder.build(),
+            true,
+            indicesRequest.indices()
+        );
+
+        Map<String, IndexMetadata> indexMetadataMap = new HashMap<>();
+        for (Index index : indices) {
+            IndexService indexService = indexServiceSafe(index);
```

Comment:
Using `indexServiceSafe` here creates CI test failures. Investigating these failures locally, their nature is inconsistent and doing things like setting interactive breakpoints and adding static waits after index creation intermittently fixes the failures. Smells like a race condition...

Reply:
I think I figured it out: `IndicesService` is built per node and the `indices` map only contains the indices that exist on that particular node. The tests that are failing use multiple nodes, so the failures are random based on node-index assignment.